Integrating angular-key-value-editor directive for build configs...

rhamilto opened the pull request:
...and deployment configs.  Also updated the Environment tabs of build, deployment, and pod pages to include "Environment Variables" in the heading.

@benjaminapetersen, @jwforres, PTAL.

benjaminapetersen commented on app/scripts/app.js:
```diff
@@ -227,7 +228,7 @@ angular
         templateUrl: 'views/other-resources.html',
         controller: 'OtherResourcesController'
-      })      
+      })
```
We should possibly double check and see if everyone has the same settings for eliminating extra white space....

benjaminapetersen commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -65,6 +66,46 @@ angular.module('openshiftConsole')
+              $scope.buildStrategy.env = [];
+            }
+            $scope.envVars = $scope.buildStrategy.env;
+
```
I think some of these comments can be trimmed.

benjaminapetersen commented on bower.json:
```diff
@@ -41,7 +41,8 @@
     "ui-select": "angular-ui-select#0.16.1",
-    "matchHeight": "0.6.0"
+    "matchHeight": "0.6.0",
+    "angular-key-value-editor": "https://github.com/benjaminapetersen/angular-key-value-editor.git#v0.0.9"
```
Just making a note on our conversation about pulling `angular-key-value-editor` under the `openshift` org temporarily, then possibly to `patternfly` as a better long term resting place.

jwforres commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -65,6 +66,40 @@ angular.module('openshiftConsole')
             }
 
+            $scope.updatedBuildConfig = angular.copy($scope.buildConfig);
+            $scope.buildStrategy = $filter('buildStrategy')($scope.updatedBuildConfig);
```
dont have a local variable and a scope variable that use the same name and are not pointing to the same thing

jwforres replied:
actually, does this even need to be on the scope object, the only references to buildStrategy in the views that i am seeing are using the filter, not this object

jwforres commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -65,6 +66,40 @@ angular.module('openshiftConsole')
 
+            $scope.updatedBuildConfig = angular.copy($scope.buildConfig);
+            $scope.buildStrategy = $filter('buildStrategy')($scope.updatedBuildConfig);
+            $scope.strategyType = $scope.buildConfig.spec.strategy.type;
```
doesn't look like this is being used anywhere

jwforres commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -65,6 +66,40 @@ angular.module('openshiftConsole')
+
+            $scope.saveEnvVars = function() {
+              $scope.envVars = _.reject(  $scope.envVars, function(env) {
+                return env.name === "";
```
any reason not to have `return !env.name`  that way if anything ever changed and you got back null / undefined name then it would still reject it

jwforres commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -65,6 +66,40 @@ angular.module('openshiftConsole')
+                }, function error(){
+                  $scope.alerts['saveBCEnvVarsError'] = {
+                    type: "error",
+                    // TODO:  improve error alert
```
we have a helper filter already for extracting failures from the response, go ahead and implement this TODO please

benjaminapetersen replied:
`details: "Reason: " + $filter('getErrorDetails')(e)`

benjaminapetersen replied:
```javascript
 $scope.alerts['saveBCEnvVarsError'] = {
  type: "error",
  message: $scope.buildConfigName + " was not updated."
  details: "Reason: " + $filter('getErrorDetails')(e)
}

```

jwforres commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -74,6 +109,7 @@ angular.module('openshiftConsole')
                 };
               }
               $scope.buildConfig = buildConfig;
+              $scope.updatedBuildConfig = angular.copy(buildConfig);
```
if anything updates the build config it could wipe out any changes you've made in the form, background controllers could potentially trigger an update to the BC's status (example, the latest build number that was run)

benjaminapetersen replied:
Seems like the right thing to do might be to ignore the watch while editing and let the "submit" fail because then the user would know that someone else edited/changed it already?

spadgett replied:
That's what we do elsewhere. It would be nice to give the user a warning though before they go through the trouble editing.

benjaminapetersen replied:
Yeah, easy would be a banner "This resource has been updated. refresh?" and let the user decide if it should replace what they have.  Harder is the diff...

@rhamilto just to keep you in the loop on this chatter :)

rhamilto replied:
In addressing the other comments here, I noticed that changes to env vars made via `Edit YAML` are not reflected in the key-value-editor.  I assume this aforementioned bug is related to this same issue.

jwforres commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -63,6 +63,17 @@ angular.module('openshiftConsole')
+    // copy deploymentConfig and ensure it has env so that we can edit env vars using key-value-editor
+    var copyDeploymentConfigAndEnsureEnv = function(deploymentConfig) {
+      $scope.updatedDeploymentConfig = angular.copy(deploymentConfig);
+      $scope.updatedDeploymentConfig.spec.template.spec.containers = _.map($scope.updatedDeploymentConfig.spec.template.spec.containers, function(container){
```
wouldn't this be easier to read as
```
_.each($scope.updatedDeploymentConfig.spec.template.spec.containers, function(container) {
  container.env = container.env || [];
}
```
updatedDeploymentConfig is already a copy, i don't see the need for creating a new array (which i assume was the reason for using map in the first place)?  @benjaminapetersen  ?

benjaminapetersen replied:
The `_.each` version is definitely shorter!  I think this was just a decision of favoring `map` over `each`, but I agree, each is a good call here.

jwforres commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -87,6 +98,34 @@ angular.module('openshiftConsole')
+              var updatedDeploymentConfig = angular.copy($scope.updatedDeploymentConfig);
+              updatedDeploymentConfig.spec.template.spec.containers = _.map(updatedDeploymentConfig.spec.template.spec.containers, function(container) {
+                container.env = _.reject(container.env, function(env){
+                  return env.name === "" && env.value ==="";
```
same comment as before,  `!env.name && !env.value`

jwforres replied:
actually just `!env.name` this never got updated to only reject when the name is empty, you may want to set an env to an empty value

jwforres commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -87,6 +98,34 @@ angular.module('openshiftConsole')
+            $scope.saveEnvVars = function() {
+              // copy again to clean off any of angulars $$ vars from the template
+              var updatedDeploymentConfig = angular.copy($scope.updatedDeploymentConfig);
+              updatedDeploymentConfig.spec.template.spec.containers = _.map(updatedDeploymentConfig.spec.template.spec.containers, function(container) {
```
same question on _.map vs just _.each

jwforres commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -87,6 +98,34 @@ angular.module('openshiftConsole')
+                }, function error(){
+                  $scope.alerts['saveDCEnvVarsError'] = {
+                    type: "error",
+                    // TODO:  improve error alert
```
implement error message

jwforres commented on app/styles/main.less:
```diff
@@ -81,6 +81,9 @@
 @import "@{pf-less-path}/vertical-nav.less";
 
+/* Angular Key Value Editor */
+@import "../../bower_components/angular-key-value-editor/src/less/angular-key-value-editor.less";
```
would prefer to keep all the relative path calculations in variables at the very top of the file, create a new var for bower-components-path

rhamilto commented:
@jwforres, I believe I've addressed all comments.  

Given that this only adds the new key-value-editor to the environment tab of build configs and deployment configs, do we want to wait to merge until we've also swapped the instances of osc-key-values for key-value-editor, too?  Or do we want a separate story and PR?  We sort of blew up the scope of the original story...

benjaminapetersen commented:
I def like separate PRs for chunks of this rather than carrying one big PR, but can also see the concern of having both editors existing during a transition period.  Curious your thoughts as well @jwforres

jwforres commented:
im ok with separate PRs to keep the review scope down but I would prefer to
not complete the card till we have them all updated.  or add a new card to
update the create form.

On Wed, Jul 6, 2016 at 10:50 AM, Ben Petersen <notifications@github.com>
wrote:

> I def like separate PRs for chunks of this rather than carrying one big
> PR, but can also see the concern of having both editors existing during a
> transition period. Curious your thoughts as well @jwforres
> <https://github.com/jwforres>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/121#issuecomment-230796058>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABZk7To0QoRrscr9l09f6Uo4qE2oEQEEks5qS8CxgaJpZM4JBixZ>
> .
>

benjaminapetersen commented:
@rhamilto I see the `angular-key-value-editor` entry in `bower.json` is registered now.  Looks like this might be good?

rhamilto commented:
@benjaminapetersen, I think so, but the build is failing.  And it's not clear to me why.

benjaminapetersen commented:
aye, taking a look.

benjaminapetersen commented:
line 915 in the log claims its not available:
```
  Error: [$injector:nomod] Module 'key-value-editor' is not available! You either misspelled the module name or forgot to load it. If registering a module ensure that you specify the dependencies as the second argument.
```
hmm.

benjaminapetersen commented:
I see it in the index.html, app.js and dist files.  Still looking.

benjaminapetersen commented:
Yup, running `grunt test` locally will result in the same problems.

benjaminapetersen commented:
@rhamilto, the `test/karma.conf.js` file just needs to be updated and add the main file for `angular-key-value-editor` at ~ line 55:
```javascript
// lots of stuff above
"bower_components/angular-extension-registry/dist/angular-extension-registry.min.js",
"bower_components/angular-key-value-editor/dist/angular-key-value-editor.min.js", // <-- add this line
"bower_components/ng-sortable/dist/ng-sortable.js",
// lots of stuff below
```
tests should run and pass then. you can give it a spin locally with `grunt test`, but it ran for me.

rhamilto commented:
@benjaminapetersen, thank you!  Success!  @jwforres or @spadgett. ready for merge.

spadgett commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -87,6 +95,33 @@ angular.module('openshiftConsole')
             ImageStreamResolver.fetchReferencedImageStreamImages([deploymentConfig.spec.template], $scope.imagesByDockerReference, $scope.imageStreamImageRefByDockerReference, context);
+            copyDeploymentConfigAndEnsureEnv(deploymentConfig);
+            $scope.saveEnvVars = function() {
+              // copy again to clean off any of angulars $$ vars from the template
```
Is this needed? Angular toJSON strips these, which I'd guess is called under the hood when we make the HTTP request.

https://docs.angularjs.org/api/ng/function/angular.toJson

rhamilto replied:
@benjaminapetersen?

benjaminapetersen replied:
Might have been overly cautious doing the extra copy w/o first trying it.  I'd say remove & test it. If it still works, lets drop the extra angular.copy().  

https://github.com/angular/angular.js/blob/master/src/ng/http.js
Looks like `$http` uses `.toJSON` (via `.serializeValue`)

spadgett commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -87,6 +95,33 @@ angular.module('openshiftConsole')
+              _.each(updatedDeploymentConfig.spec.template.spec.containers, function(container) {
+                container.env = _.reject(container.env, function(env){
+                  return !env.name;
+                });
```
FYI you should be able to use

```js
container.env = _.filter(container.env, 'name');
```

rhamilto replied:
Neat!  Fixed (commit forthcoming).

spadgett commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -42,8 +42,15 @@ angular.module('openshiftConsole')
+    // copy buildConfig and ensure it has env so that we can edit env vars using key-value-editor
+    var copyBuildConfigAndEnsureEnv = function(buildConfig) {
+      $scope.updatedBuildConfig = angular.copy(buildConfig);
+      $scope.envVars = $filter('buildStrategy')($scope.updatedBuildConfig).env || [];
```
If you move the func after the `buildStrategy` declaration just below, you won't have the call `$filter`.

```js
$scope.envVars = buildStrategy($scope.updatedBuildConfig).env || [];
```

rhamilto replied:
Good catch.  Thanks.

spadgett commented on app/scripts/controllers/deploymentConfig.js:
```diff
@@ -63,6 +63,14 @@ angular.module('openshiftConsole')
+      $scope.updatedDeploymentConfig = angular.copy(deploymentConfig);
+      _.each($scope.updatedDeploymentConfig.spec.template.spec.containers, function(container) {
+        container.env = container.env || [];
+      });
```
Same comment as below.

rhamilto replied:
As below?  Sorry, I'm not clear on what you're getting at here.

spadgett replied:
Sorry disregard

rhamilto commented:
@spadgett, thanks for the suggestions.  I've integrated them and think we're now ready for merge.  *crosses fingers*

spadgett commented:
Can be a follow-on, but I'd like to see a warning if you navigate away without saving. It'd be easy to miss the save button. Maybe also a reset button.

spadgett commented on app/scripts/controllers/buildConfig.js:
```diff
@@ -66,6 +73,32 @@ angular.module('openshiftConsole')
+              buildStrategy($scope.updatedBuildConfig).env = $scope.envVars;
+
+              DataService
+                .update("buildconfigs", $routeParams.buildconfig, angular.copy($scope.updatedBuildConfig), context)
```
Sorry, just noticed: do we need to copy again here? `updatedBuildConfig` is already a copy, right?

spadgett commented:
I really like the way the new editor works.

spadgett commented:
One last question on angular.copy, otherwise LGTM.

rhamilto commented:
> One last question on angular.copy, otherwise LGTM.

Another good catch, @spadgett.  It wasn't needed, so I removed it.

benjaminapetersen commented:
Nice, looking forward to this landing!

spadgett commented:
@rhamilto ready to merge?

rhamilto commented:
👍

spadgett commented:
[merge]
